controller/composite/controller.go

```diff
@@ -434,20 +434,34 @@ func (pc *parentController) syncParentObject(parent *unstructured.Unstructured)
 		parent = updatedParent
 	}
 
-	// If selector generation is enabled, add the controller-uid label to all
-	// desired children.
-	if pc.cc.Spec.GenerateSelector != nil && *pc.cc.Spec.GenerateSelector {
-		for _, group := range desiredChildren {
-			for _, obj := range group {
-				labels := obj.GetLabels()
-				if labels == nil {
-					labels = make(map[string]string, 1)
+	// Enforce invariants between parent selector and child labels.
+	selector, err := pc.makeSelector(parent, nil)
+	if err != nil {
+		return err
+	}
+	for _, group := range desiredChildren {
+		for _, obj := range group {
+			// We don't use GetLabels() because that swallows conversion errors.
+			objLabels, _, err := unstructured.NestedStringMap(obj.UnstructuredContent(), "metadata", "labels")
+			if err != nil {
+				return fmt.Errorf("invalid labels on desired child %v %v/%v: %v", obj.GetKind(), obj.GetNamespace(), obj.GetName(), err)
+			}
+			// If selector generation is enabled, add the controller-uid label to all
+			// desired children.
```

Comment:
Suggestion: add more details on why `controller-uid` label is needed when selector generation is enabled.

Reply:
Done.